src/mlpack/methods/amf/init_rules/average_init.hpp

```diff
@@ -70,8 +68,57 @@ class AverageInitialization
     W.randu(n, r);
     H.randu(r, m);
 
-    W = W + avgV;
-    H = H + avgV;
+    W += avgV;
+    H += + avgV;
+  }
+
+  /**
+   * Initialize the matrix W or H to the average value of V with uniform
+   * random noise added.
+   *
+   * @param V Input matrix.
+   * @param r Rank of matrix.
+   * @param M W or H matrix, to be initialized to the average value of V
+   * with uniform random noise added.
+   * @param whichMatrix If true, initialize W. Otherwise, initialize H.
+   */
+  template<typename MatType>
+  inline static void InitializeOne(const MatType& V,
+                                   const size_t r,
+                                   arma::mat& M,
+                                   const bool whichMatrix = true)
+  {
+    const size_t n = V.n_rows;
+    const size_t m = V.n_cols;
+
+    double avgV = 0;
+    double min = DBL_MAX;
+
+    // Iterate over all elements in the matrix (for sparse matrices, this only
+    // iterates over nonzeros).
+    for (typename MatType::const_row_col_iterator it = V.begin();
+        it != V.end(); ++it)
+    {
+      avgV += *it;
+      // Track the minimum value.
+      if (*it < min)
+        min = *it;
+    }
+
+    avgV = sqrt(((avgV / (n * m)) - min) / r);
```

Comment:
I guess there is no need to create a copy if we only use this once? See comment below.

Reply:
Hi @zoq, what do you mean by "no need to create a copy"?

Reply:
You can directly use: `M += sqrt(((avgV / (n * m)) - min) / r);` instead of writing it to `avgV` first.

Reply:
Got it! Thanks for pointing this out.